### ISSUE TITLE
Tons of "possible loss of data" and other warnings on x64

### ISSUE THREAD

**Issue by rustyx** [jemalloc/jemalloc#317]
On Windows x64, a lot of "possible loss of data" warnings is logged. ([link](http://rustyx.org/temp/jemalloc_msvc_warnings.txt))

Could it be because jemalloc incorrectly assumes `sizeof(void*)` == `sizeof(long)` which does not hold on this platform?

```
checking size of void *... 8
checking size of int... 4
checking size of long... 4
checking size of intmax_t... 8
```

BTW similar warnings can be reproduced in FreeBSD amd64 if compiled with `-Wshorten-64-to-32`. ([link](http://rustyx.org/temp/jemalloc_gcc_warnings.txt))

**Comment by jasone** [comment]
I doubt that any of these warnings indicate actual problems, though it's probably worth explicitly casting where appropriate to remove the warnings so that we can use future warnings to help prevent bugs.

**Comment by jasone** [comment]
After having been bitten by several bad assumptions re: ```sizeof(long) != sizeof(void *)```, I'm inclined to think this is worth digging into.

**Comment by rustyx** [comment]
A good data type for working with pointers is `uintptr_t` (from stdint.h). Is optional in C99, but seems to be defined in GCC, clang, Mingw and MSVC (since 2010), so quite widely available IMO.
The "long" functions like `ffsl` etc. should then operate on this data type.

It would also be nice to define a custom macro for explicit 64-to-32 conversion, e.g. SHORTEN_TO_INT() or something like it. That makes places of potential 32-bit overflow more visible.

**Comment by jasone** [comment]
jemalloc uses ```uintptr_t``` extensively (anytime explicit pointer arithmetic is required).  The uses of ```long``` are limited to interactions with e.g. ```ffsl()```, atomic operations, etc..  I went through a round of portability bugs over the past week due to incorrect conversions interacting with ```struct timespec```, which prompted the introduction of ```jemalloc_ffs64()```.

Today I'm going to follow your suggestion of using ```-Wshorten-64-to-32``` and work through the issues that uncovers.  However, a quick ```git grep long``` suggests that atomic operations are the only likely remaining issue area.

**Comment by yuslepukhin** [comment]
Use of size_t may be appropriate in many places instead of unsigned int. This would make it uniform across the platforms and change its size automatically depending on the architecture.


-------- Original Message --------

**Comment by jasone** [comment]
This is resolved by 9f4ee6034c3ac6a8c8b5f9a0d76822fb2fd90c41, 8dd5115edee9e778d3b45d0924530ee49a4e34e6, 603b3bd413d670909811ce49d28a8b0a4ae3ba6b, 8f683b94a751c65af8f9fa25970ccf2917b96bb8, 0931cecbfaeada8b10fed56ff0175c8ffb9e9233, 1c42a04cc6d3cc5d92bec55432015785584a4b0d, 9e1810ca9dc4a5f5f0841b9a6c1abb4337753552, b3d0070b1495ddd36893d481c512b5da1ab8acef, and ca8fffb5c13b6a7c45fd034667a8910c61d09c3b.  I tested with clang on OS X, Linux, and FreeBSD.

**Comment by jasone** [comment]
aa63d5d377b4508b83502e923690d1d7b67c8c88 fixes a compilation error on MinGW.
